xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/PVRClient.h

```diff
@@ -0,0 +1,4763 @@
+/*
+ *  Copyright (C) 2005-2018 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#pragma once
+
+#include "../AddonBase.h"
+
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <time.h>
+
+/*! @note Define "USE_DEMUX" at compile time if demuxing in the PVR add-on is used.
+ *        Also, "DVDDemuxPacket.h" file must be in the include path of the add-on,
+ *        and the add-on should set bHandlesDemuxing to true.
+ */
+#ifdef USE_DEMUX
+#include "../DemuxPacket.h"
+#else
+struct DemuxPacket;
+#endif
+
+//==============================================================================
+/// \defgroup cpp_kodi_addon_pvr_Defs Definitions, structures and enumerators
+/// \ingroup cpp_kodi_addon_pvr
+/// @brief **PVR client add-on instance definition values**
+//------------------------------------------------------------------------------
+
+#define PVR_ADDON_NAME_STRING_LENGTH 1024
+#define PVR_ADDON_URL_STRING_LENGTH 1024
+#define PVR_ADDON_DESC_STRING_LENGTH 1024
+#define PVR_ADDON_INPUT_FORMAT_STRING_LENGTH 32
+#define PVR_ADDON_EDL_LENGTH 32
+#define PVR_ADDON_TIMERTYPE_ARRAY_SIZE 32
+#define PVR_ADDON_TIMERTYPE_VALUES_ARRAY_SIZE 512
+#define PVR_ADDON_TIMERTYPE_VALUES_ARRAY_SIZE_SMALL 128
+#define PVR_ADDON_TIMERTYPE_STRING_LENGTH 128
+#define PVR_ADDON_ATTRIBUTE_DESC_LENGTH 128
+#define PVR_ADDON_ATTRIBUTE_VALUES_ARRAY_SIZE 512
+#define PVR_ADDON_DESCRAMBLE_INFO_STRING_LENGTH 64
+
+/* defines for GetChannelStreamProperties, GetRecordingStreamProperties and GetEPGTagStreamProperties */
+#define PVR_STREAM_MAX_PROPERTIES 20
+
+/* using the default avformat's MAX_STREAMS value to be safe */
+#define PVR_STREAM_MAX_STREAMS 20
+
+namespace kodi
+{
+namespace addon
+{
+  class CInstancePVRClient;
+} /* namespace addon */
+} /* namespace kodi */
+
+extern "C"
```

Comment:
Also, having extern "C"inside a namespace (which is C++) feels wrong.

We should move C-API definition out of the kodi::addon namespace.

Reply:
It's not inside a namespace, see two line above (59) there is it ended.

To have this C and counterpart C++ structure near together is mostly to make it easier on changes. A extern "C" does only say the compiler that there parts where nedded on "C" code and it prevent some special "C++" , there good to prevent some mistakes.

Reply:
Not sure it is good design to mix  C and C++ code in  an  'extern "C"' block. I would never do that.